### ISSUE TITLE
wxStyledTextCtrl isn't accessible with nvda screenreader

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#18683]
*Issue migrated from trac ticket # 18683*

**component:** wxStyledText | **priority:** normal | **keywords:** accessibility

#### 2020-02-26 12:57:18: gumerov-amir (amir) created the issue

___
styledTextCtrl isn't accessable with nvda screenreader

**Comment by hwf1324** [comment]
The current stc sample built on the master branch should be resolved https://github.com/nvaccess/nvda/issues/13976

But the above problem still exists in the 3.2.4 build. I'm not sure if the update to Scintilla solved this issue at the same time?

Also not sure if there are other accessibility issues?

Are there any plans to port this to 3.2.x?

**Comment by MaartenBent** [comment]
Scintilla can't be backported to 3.2. Because 3.2 is compatible with c++98, and Scintilla in the master branch requires c++11.

But maybe only the relevant changes can be backported. I'll try to install nvda and verify that it works in master. And then try to figure out what fixed it. One of the Scintilla updates, or something in wxWidgets.

**Comment by MaartenBent** [comment]
Seems like a bug in wxWidgets, not Scintilla. I compiled master with Scintilla 3.7.2 (the one used in 3.2 branch) and it works fine. When I run the stc sample of the 3.2 branch, the problem appears.

For reference, the problem from the linked issue:

```
Write some text ex. "Hello world" on the first row then press Enter
On the second row write "Hi"

Actual behavior:
In NVDA braille viewer the content with the cursor on the second row is:
" Hi llo world"

It seems in general that after the end of the row it concatenates the remaining characters from the longest row of the control.
```

I'll try to bisect and find out what fixed it.

**Comment by hwf1324** [comment]
Thank you very much for your work.

**Comment by MaartenBent** [comment]
It was fixed by fcc7430610ca202b744df75dc3d7e471a1e8b8a4 `Set wxALWAYS_NATIVE_DOUBLE_BUFFER to 1 to wxMSW too`.

I don't think we can just apply that to 3.2, because it assumes all other double buffering and composition changes are applied.

The relevant line in the stc code is:
https://github.com/wxWidgets/wxWidgets/blob/b97aee419af105e5c724ead2917e1b04b9abb26a/src/stc/stc.cpp#L228-L234

Where `SetBufferedDraw(false);` fixes it. But this causes flickering without the double buffering fixes from master.

I'm not sure how NVDA works, but it seems to do some stuff with `SCI_` messages. Maybe they can disable buffered drawing with the `SCI_SETBUFFEREDDRAW` message for `stcwindow` windows?

**Comment by hwf1324** [comment]
I don't really understand this aspect. Maybe this is indeed a temporary way?

Cc @lukaszgo1, @seanbudd, @michaelDCurran Any thoughts?

**Comment by MaartenBent** [comment]
Maybe something else to look at. Is the `wxStyledTextCtrl` correctly identified as Scintilla control?
Spy++ says the classname of `wxStyledTextCtrl` will be `wxWindow` with window caption `stcwindow`.

https://github.com/nvaccess/nvda/blob/6d97c5a8b74bc423c13a218b7575b2b7be8858e9/source/NVDAObjects/window/__init__.py#L121-L122
```
elif windowClassName in ("Scintilla","TScintilla"):
    from .scintilla import Scintilla as newCls
```

**Comment by hwf1324** [comment]
> It was fixed by [fcc7430](https://github.com/wxWidgets/wxWidgets/commit/fcc7430610ca202b744df75dc3d7e471a1e8b8a4) `Set wxALWAYS_NATIVE_DOUBLE_BUFFER to 1 to wxMSW too`.
> 
> I don't think we can just apply that to 3.2, because it assumes all other double buffering and composition changes are applied.
> 
> The relevant line in the stc code is:
> 
> https://github.com/wxWidgets/wxWidgets/blob/b97aee419af105e5c724ead2917e1b04b9abb26a/src/stc/stc.cpp#L228-L234
> 
> Where `SetBufferedDraw(false);` fixes it. But this causes flickering without the double buffering fixes from master.
> 
> I'm not sure how NVDA works, but it seems to do some stuff with `SCI_` messages. Maybe they can disable buffered drawing with the `SCI_SETBUFFEREDDRAW` message for `stcwindow` windows?

I followed [this article](http://www.zlotowicz.pl/nvda/winapi.mdwn) and sent a window message to the `wx.StyledTextCtrl` control, but it didn't work.

NVDA has a gesture that calls out the wxPython GUI inspection tool. It contains the `wx.StyledTextCtrl` control. StyledTextCtrl` control, which can be found in the NVDA menu -> Preferences -> Input gestures... Input gestures... Under the Tools category.

Press `NVDA+ctrl+z` to open the Python console after moving focus to this control.

```Python
>>> import watchdog
>>> SCI_GETBUFFEREDDRAW = 2034
>>> SCI_SETBUFFEREDDRAW = 2035
>>> nav.name
'stcwindow'
>>> nav.windowClassName
'wxWindowNR'
>>> watchdog.cancellableSendMessage(nav.windowHandle, SCI_GETBUFFEREDDRAW, 0, 0)
0
>>> watchdog.cancellableSendMessage(nav.windowHandle, SCI_SETBUFFEREDDRAW, 1, 0)
0
>>> watchdog.cancellableSendMessage(nav.windowHandle, SCI_GETBUFFEREDDRAW, 0, 0)
0
>>> watchdog.cancellableSendMessage(nav.windowHandle, SCI_SETBUFFEREDDRAW, 0, 0)
0
>>> watchdog.cancellableSendMessage(nav.windowHandle, SCI_GETBUFFEREDDRAW, 0, 0)
0
```

**Comment by Adriani90** [comment]
@MaartenBent any updates on this from wX side? Are you saying this should be fixed on NVDA side?

**Comment by MaartenBent** [comment]
I looked at it again, and I think I figured out the problems. NVDA doesn't recognize our control as Scintilla control, so it won't send scintilla messages (`SCI_...`) to get text and stuff.
I can fix that by using `Scintilla` as window class instead of `wxWindowNR`.

Second problem is we don't forward the scintilla messages to the Scintilla control. This is easy to fix.
We already forward `WM_IME_` messages. I have to extend this to include the `SCI_` messages.

With these fixes, all the lines appear correctly in the speech viewer. I'll try to create a PR with these fixes this weekend.

The double buffering stuff was a red herring. Because it wasn't recognized as Scintilla control, it was handled as normal Edit control instead. And this doesn't seem to work nicely with our double buffering.
